### DISCUSSION TITLE
Custom flag aligning (padding)

### DISCUSSION THREAD

**Discussion by agil899** [gabime/spdlog#2398] [Q&A] [answered]
I've implemented my own custom flag `y` following the README guidelines and I'd like to add aligning so that the length of the text is constant.
I tried `%-10y` but this doesn't work, is there a different way to do it for custom flags?

**Comment by tt4g** [comment, marked as answer]
Adding padding seems to be the role of the formatter.
As far as I can tell from the source of the `short_level_formatter`, it passes `padding_info` to `ScopedPadder` to add padding.
The `ScopedPadder` is passed as a template argument, either `scoped_padder` or `null_scoped_padder`.

https://github.com/gabime/spdlog/blob/76fb40d95455f249bd70824ecfcae7a8f0930fa3/include/spdlog/pattern_formatter-inl.h#L138-L153

https://github.com/gabime/spdlog/blob/76fb40d95455f249bd70824ecfcae7a8f0930fa3/include/spdlog/pattern_formatter-inl.h#L1397-L1404

For custom formatters, it appears that `set_padding_info()` is called and `padding_info` is passed to the `padinfo_` member variable.

https://github.com/gabime/spdlog/blob/76fb40d95455f249bd70824ecfcae7a8f0930fa3/include/spdlog/pattern_formatter-inl.h#L1096-L1104

https://github.com/gabime/spdlog/blob/76fb40d95455f249bd70824ecfcae7a8f0930fa3/include/spdlog/pattern_formatter.h#L71-L75

So adding padding in the custom formatting function should work.

Example (NOTE has not been tested):

```cpp
#include "spdlog/pattern_formatter.h"
class my_formatter_flag : public spdlog::custom_flag_formatter
{
public:
    void format(const spdlog::details::log_msg &, const std::tm &, spdlog::memory_buf_t &dest) override
    {
        std::string some_txt = "custom-flag";
        if (this->padinfo_.enabled()) 
        { 
            spdlog::details::scoped_padder(some_txt.size(), this->padinfo_, dest); 
        }
        dest.append(some_txt.data(), some_txt.data() + some_txt.size());
    }

    std::unique_ptr<custom_flag_formatter> clone() const override
    {
        return spdlog::details::make_unique<my_formatter_flag>();
    }
};
```

**Reply by agil899** [reply]
Thanks a lot for your help!
You were right except for a small detail that `dest.append()` needs to be in the same scope as `scopped_padder` for left aligning to work. Also there needs to be an instance of `scopped_padder` so that the destructor doesn't get called immediately.
Final result:
```c++
#include "spdlog/pattern_formatter.h"
class my_formatter_flag : public spdlog::custom_flag_formatter
{
public:
    void format(const spdlog::details::log_msg &, const std::tm &, spdlog::memory_buf_t &dest) override
    {
        std::string some_txt = "custom-flag";
        if (this->padinfo_.enabled()) 
        { 
            spdlog::details::scoped_padder p(some_txt.size(), this->padinfo_, dest); 
            dest.append(some_txt.data(), some_txt.data() + some_txt.size());
            return;
        }
        dest.append(some_txt.data(), some_txt.data() + some_txt.size());
    }

    std::unique_ptr<custom_flag_formatter> clone() const override
    {
        return spdlog::details::make_unique<my_formatter_flag>();
    }
};
```
